### PR TITLE
Upgrade SDK to v2

### PR DESCRIPTION
**Pull request by hahattan**
Sorry for the inconvenience but  it's really hard to partially refactor the SDK to v2 without building error.
I try to narrow down the changes in each commit and hope it helps the review.
Most of the changes are just updating the `go-mod-core-contracts` path to use v2 models.
For some massive change I've added the explanation in the commit body.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-sdk-c/blob/master/.github/CONTRIBUTING.md

## What is the new behavior?
SDK is compatible with v2 models/API
- The `ProtocolDriver` implementation, the external function API in `pkg/service/managed*.go` needs to change to use v2 models.
- The `/api/v1` route will no longer function.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

## New Imports

- [ ] Yes
- [x] No

## Other information

as discussed in #746, the `compareReadings` needs to change back to use hash. Also this PR doesn't handle the CBOR encoding for pushing binary event because the `EventClient` doesn't have the support.

I think it's better to update it in another PR

### CODE REVIEW

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]

File: pkg/service/async.go [hunk 1]
```diff
@@ -133,28 +87,32 @@ func (s *DeviceService) processAsyncFilterAndAdd(ctx context.Context, wg *sync.W
...
-							Service:        pw.Service,
+							ServiceName:    pw.ProfileName,
 							AdminState:     pw.AdminState,
-							OperatingState: contract.Enabled,
+							OperatingState: contract.Up,
 							AutoEvents:     nil,
 						}
-						device.Origin = millis
-						device.Description = d.Description
+						req := requests.AddDeviceRequest{
```
Comment (cloudxxx8) [inline pkg/service/async.go:106]: use `NewAddDeviceRequest` when it is available

File: pkg/service/managedautoevents.go [hunk 2]
```diff
@@ -35,35 +36,43 @@ func (s *DeviceService) AddDeviceAutoEvent(deviceName string, event contract.Aut
...
 	device, ok := cache.Devices().ForName(deviceName)
 	if !ok {
-		msg := fmt.Sprintf("Device %s cannot be found in cache", deviceName)
+		msg := fmt.Sprintf("failed to find device %s cannot in cache", deviceName)
 		s.LoggingClient.Error(msg)
 		return fmt.Errorf(msg)
 	}
 
-	autoevent.GetManager().StopForDevice(deviceName)
+	s.manager.StopForDevice(deviceName)
```
Comment (cloudxxx8) [inline pkg/service/managedautoevents.go:61]: do we need to stop it here?
it is restarted in the end of the function.

File: pkg/service/manageddevices.go [hunk 3]
```diff
@@ -10,125 +10,108 @@ package service
...
 		}
 	}
 
-	millis := time.Now().UnixNano() / int64(time.Millisecond)
-	device.Origin = millis
-	device.Service = *s.deviceService
-	device.Profile = prf
-
+	s.LoggingClient.Debugf("Adding managed Device %s", device.Name)
+	req := requests.AddDeviceRequest{
```
Comment (cloudxxx8) [inline pkg/service/manageddevices.go:42]: use `NewAddDeviceRequest` func when it is available

File: pkg/service/manageddevices.go [hunk 4]
```diff
@@ -10,125 +10,108 @@ package service
...
-		msg := fmt.Sprintf("Device %s cannot be found in cache", device.Id)
+		msg := fmt.Sprintf("failed to find Device %s in cache", device.Name)
 		s.LoggingClient.Error(msg)
 		return fmt.Errorf(msg)
 	}
 
-	s.LoggingClient.Debug(fmt.Sprintf("Updating managed Device: : %s\n", device.Name))
+	s.LoggingClient.Debugf("Updating managed Device %s", device.Name)
+	req := requests.UpdateDeviceRequest{
+		BaseRequest: commonDTO.BaseRequest{
```
Comment (cloudxxx8) [inline pkg/service/manageddevices.go:106]: use NewBaseRequest func

File: pkg/service/manageddevices.go [hunk 5]
```diff
@@ -139,16 +122,25 @@ func (s *DeviceService) UpdateDevice(device contract.Device) error {
...
-		msg := fmt.Sprintf("Device %s cannot be found in cache", deviceName)
+		msg := fmt.Sprintf("failed to find Device %s in cache", deviceName)
 		s.LoggingClient.Error(msg)
 		return fmt.Errorf(msg)
 	}
 
-	s.LoggingClient.Debug(fmt.Sprintf("Updating managed Device OperatingState: : %s\n", d.Name))
+	s.LoggingClient.Debugf("Updating managed Device OperatingState %s", d.Name)
+	req := requests.UpdateDeviceRequest{
+		BaseRequest: commonDTO.BaseRequest{
```
Comment (cloudxxx8) [inline pkg/service/manageddevices.go:132]: use NewBaseRequest func

File: pkg/service/managedprofiles.go [hunk 6]
```diff
@@ -10,133 +10,110 @@ package service
...
 	if p, ok := cache.Profiles().ForName(profile.Name); ok {
 		return p.Id, fmt.Errorf("name conflicted, Profile %s exists", profile.Name)
 	}
 
-	s.LoggingClient.Debug(fmt.Sprintf("Adding managed Profile: : %s", profile.Name))
-	millis := time.Now().UnixNano() / int64(time.Millisecond)
-	profile.Origin = millis
-
+	s.LoggingClient.Debugf("Adding managed Profile %s", profile.Name)
+	req := requests.DeviceProfileRequest{
```
Comment (cloudxxx8) [inline pkg/service/managedprofiles.go:32]: use NewDeviceProfileRequest func

File: pkg/service/managedprofiles.go [hunk 7]
```diff
@@ -10,133 +10,110 @@ package service
...
 	"github.com/google/uuid"
+
+	"github.com/edgexfoundry/device-sdk-go/v2/internal/common"
+	"github.com/edgexfoundry/device-sdk-go/v2/internal/v2/cache"
 )
 
 // AddDeviceProfile adds a new DeviceProfile to the Device Service and Core Metadata
 // Returns new DeviceProfile id or non-nil error.
-func (s *DeviceService) AddDeviceProfile(profile contract.DeviceProfile) (id string, err error) {
+func (s *DeviceService) AddDeviceProfile(profile models.DeviceProfile) (string, error) {
```
Comment (cloudxxx8) [inline pkg/service/managedprofiles.go:26]: use errors.EdgeX instead of normal error

File: pkg/service/managedprofiles.go [hunk 8]
```diff
@@ -10,133 +10,110 @@ package service
...
-		return err
-	}
-
-	err = cache.Profiles().Remove(id)
-	return err
-}
-
 // RemoveDeviceProfileByName removes the specified DeviceProfile by name from the cache and ensures that the
 // instance in Core Metadata is also removed.
 func (s *DeviceService) RemoveDeviceProfileByName(name string) error {
```
Comment (cloudxxx8) [inline pkg/service/managedprofiles.go:60]: use errors.EdgeX instead of normal error

File: pkg/service/managedprofiles.go [hunk 9]
```diff
@@ -10,133 +10,110 @@ package service
...
-	err = cache.Profiles().RemoveByName(profile.Name)
+	err = cache.Profiles().RemoveByName(name)
 	return err
 }
 
 // UpdateDeviceProfile updates the DeviceProfile in the cache and ensures that the
 // copy in Core Metadata is also updated.
-func (s *DeviceService) UpdateDeviceProfile(profile contract.DeviceProfile) error {
-	_, ok := cache.Profiles().ForId(profile.Id)
+func (s *DeviceService) UpdateDeviceProfile(profile models.DeviceProfile) error {
```
Comment (cloudxxx8) [inline pkg/service/managedprofiles.go:82]: use errors.EdgeX instead of normal error

File: pkg/service/managedprofiles.go [hunk 10]
```diff
@@ -10,133 +10,110 @@ package service
...
 	if !ok {
-		msg := fmt.Sprintf("DeviceProfile %s cannot be found in cache", profile.Id)
+		msg := fmt.Sprintf("failed to find Profile %s in cache", profile.Name)
 		s.LoggingClient.Error(msg)
 		return fmt.Errorf(msg)
 	}
 
-	s.LoggingClient.Debug(fmt.Sprintf("Updating managed DeviceProfile: : %s", profile.Name))
+	s.LoggingClient.Debugf("Updating managed Profile %s", profile.Name)
+	req := requests.DeviceProfileRequest{
```
Comment (cloudxxx8) [inline pkg/service/managedprofiles.go:91]: use NewDeviceProfileRequest

File: pkg/service/manageddevices.go [hunk 11]
```diff
@@ -10,125 +10,108 @@ package service
...
 	"github.com/google/uuid"
+
+	"github.com/edgexfoundry/device-sdk-go/v2/internal/common"
+	"github.com/edgexfoundry/device-sdk-go/v2/internal/v2/cache"
 )
 
 // AddDevice adds a new Device to the Device Service and Core Metadata
 // Returns new Device id or non-nil error.
-func (s *DeviceService) AddDevice(device contract.Device) (id string, err error) {
+func (s *DeviceService) AddDevice(device models.Device) (string, error) {
```
Comment (cloudxxx8) [inline pkg/service/manageddevices.go:26]: use errors.EdgeX instead of normal error

File: pkg/service/manageddevices.go [hunk 12]
```diff
@@ -10,125 +10,108 @@ package service
...
 
 // Devices return all managed Devices from cache
-func (s *DeviceService) Devices() []contract.Device {
+func (s *DeviceService) Devices() []models.Device {
 	return cache.Devices().All()
 }
 
 // GetDeviceByName returns the Device by its name if it exists in the cache, or returns an error.
-func (s *DeviceService) GetDeviceByName(name string) (contract.Device, error) {
+func (s *DeviceService) GetDeviceByName(name string) (models.Device, error) {
```
Comment (cloudxxx8) [inline pkg/service/manageddevices.go:64]: use errors.EdgeX instead of normal error

File: pkg/service/manageddevices.go [hunk 13]
```diff
@@ -10,125 +10,108 @@ package service
...
-		s.LoggingClient.Error(fmt.Sprintf("Delete Device %s from Core Metadata failed", id))
-	}
-
-	return err
-}
-
-// RemoveDevice removes the specified Device by name from the cache and ensures that the
+// RemoveDeviceByName removes the specified Device by name from the cache and ensures that the
 // instance in Core Metadata is also removed.
 func (s *DeviceService) RemoveDeviceByName(name string) error {
```
Comment (cloudxxx8) [inline pkg/service/manageddevices.go:76]: use errors.EdgeX instead of normal error

File: pkg/service/manageddevices.go [hunk 14]
```diff
@@ -10,125 +10,108 @@ package service
...
 	}
 
 	return err
 }
 
 // UpdateDevice updates the Device in the cache and ensures that the
 // copy in Core Metadata is also updated.
-func (s *DeviceService) UpdateDevice(device contract.Device) error {
-	_, ok := cache.Devices().ForId(device.Id)
+func (s *DeviceService) UpdateDevice(device models.Device) error {
```
Comment (cloudxxx8) [inline pkg/service/manageddevices.go:96]: use errors.EdgeX instead of normal error

File: pkg/service/manageddevices.go [hunk 15]
```diff
@@ -139,16 +122,25 @@ func (s *DeviceService) UpdateDevice(device contract.Device) error {
 func (s *DeviceService) UpdateDeviceOperatingState(deviceName string, state string) error {
```
Comment (cloudxxx8) [inline pkg/service/manageddevices.go:122]: use errors.EdgeX instead of normal error

File: pkg/service/managedautoevents.go [hunk 16]
```diff
@@ -9,24 +10,24 @@ package service
...
-	"github.com/edgexfoundry/device-sdk-go/v2/internal/cache"
-	contract "github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/v2/models"
+
+	"github.com/edgexfoundry/device-sdk-go/v2/internal/v2/cache"
 )
 
 // AddDeviceAutoEvent adds a new AutoEvent to the Device with given name
-func (s *DeviceService) AddDeviceAutoEvent(deviceName string, event contract.AutoEvent) error {
+func (s *DeviceService) AddDeviceAutoEvent(deviceName string, event models.AutoEvent) error {
```
Comment (cloudxxx8) [inline pkg/service/managedautoevents.go:19]: use errors.EdgeX instead of normal error

File: pkg/service/managedautoevents.go [hunk 17]
```diff
@@ -35,35 +36,43 @@ func (s *DeviceService) AddDeviceAutoEvent(deviceName string, event contract.Aut
...
-
-	autoevent.GetManager().RestartForDevice(deviceName, nil)
+	s.manager.RestartForDevice(deviceName)
 
 	return nil
 }
 
 // RemoveDeviceAutoEvent removes an AutoEvent from the Device with given name
-func (s *DeviceService) RemoveDeviceAutoEvent(deviceName string, event contract.AutoEvent) error {
+func (s *DeviceService) RemoveDeviceAutoEvent(deviceName string, event models.AutoEvent) error {
```
Comment (cloudxxx8) [inline pkg/service/managedautoevents.go:53]: use errors.EdgeX instead of normal error

File: pkg/service/managedwatchers.go [hunk 18]
```diff
@@ -9,81 +9,79 @@ package service
...
 	"github.com/google/uuid"
+
+	"github.com/edgexfoundry/device-sdk-go/v2/internal/common"
+	"github.com/edgexfoundry/device-sdk-go/v2/internal/v2/cache"
 )
 
 // AddProvisionWatcher adds a new Watcher to the cache and Core Metadata
 // Returns new Watcher id or non-nil error.
-func (s *DeviceService) AddProvisionWatcher(watcher contract.ProvisionWatcher) (id string, err error) {
+func (s *DeviceService) AddProvisionWatcher(watcher models.ProvisionWatcher) (string, error) {
```
Comment (cloudxxx8) [inline pkg/service/managedwatchers.go:25]: use errors.EdgeX instead of normal error

File: pkg/service/manageddevices.go (hunk 11 above)
Comment (cloudxxx8) [reply pkg/service/manageddevices.go:26]: Assign the serviceName into Device

File: pkg/service/managedwatchers.go (hunk 18 above)
Comment (cloudxxx8) [reply pkg/service/managedwatchers.go:25]: Assign the serviceName into ProvisionWatcher

File: pkg/service/managedwatchers.go [hunk 19]
```diff
@@ -9,81 +9,79 @@ package service
...
 		return "", fmt.Errorf(errMsg)
 	}
 
-	millis := time.Now().UnixNano() / int64(time.Millisecond)
-	watcher.Origin = millis
-	watcher.Service = *s.deviceService
-	watcher.Profile = prf
-
+	s.LoggingClient.Debugf("Adding managed ProvisionWatcher %s", watcher.Name)
+	req := requests.AddProvisionWatcherRequest{
```
Comment (cloudxxx8) [inline pkg/service/managedwatchers.go:38]: use NewAddProvisionWatecherRequest when it is available

File: pkg/service/managedwatchers.go [hunk 20]
```diff
@@ -92,19 +90,25 @@ func (s *DeviceService) RemoveProvisionWatcher(id string) error {
...
-		msg := fmt.Sprintf("provisionwatcher %s cannot be found in cache", watcher.Id)
+		msg := fmt.Sprintf("failed to find ProvisionWatcher %s in cache", watcher.Name)
 		s.LoggingClient.Error(msg)
 		return fmt.Errorf(msg)
 	}
 
-	s.LoggingClient.Debug(fmt.Sprintf("Updating managed ProvisionWatcher: %s", watcher.Name))
+	s.LoggingClient.Debugf("Updating managed ProvisionWatcher: %s", watcher.Name)
+	req := requests.UpdateProvisionWatcherRequest{
+		BaseRequest: commonDTO.BaseRequest{
```
Comment (cloudxxx8) [inline pkg/service/managedwatchers.go:103]: use NewBaseRequest func

File: pkg/service/managedwatchers.go [hunk 21]
```diff
@@ -9,81 +9,79 @@ package service
...
 	}
 	return pw, nil
 }
 
-// RemoveProvisionWatcher removes the specified Watcher by id from the cache and ensures that the
+// RemoveProvisionWatcher removes the specified Watcher by name from the cache and ensures that the
 // instance in Core Metadata is also removed.
-func (s *DeviceService) RemoveProvisionWatcher(id string) error {
-	pw, ok := cache.ProvisionWatchers().ForId(id)
+func (s *DeviceService) RemoveProvisionWatcher(name string) error {
```
Comment (cloudxxx8) [inline pkg/service/managedwatchers.go:72]: use errors.EdgeX instead of normal error

File: pkg/service/managedwatchers.go [hunk 22]
```diff
@@ -92,19 +90,25 @@ func (s *DeviceService) RemoveProvisionWatcher(id string) error {
 
 // UpdateProvisionWatcher updates the Watcher in the cache and ensures that the
 // copy in Core Metadata is also updated.
-func (s *DeviceService) UpdateProvisionWatcher(watcher contract.ProvisionWatcher) error {
-	_, ok := cache.ProvisionWatchers().ForId(watcher.Id)
+func (s *DeviceService) UpdateProvisionWatcher(watcher models.ProvisionWatcher) error {
```
Comment (cloudxxx8) [inline pkg/service/managedwatchers.go:93]: use errors.EdgeX instead of normal error

File: pkg/service/service.go [hunk 23]
```diff
@@ -135,109 +131,64 @@ func (s *DeviceService) Stop(force bool) {
...
 	if err != nil {
-		if errsc, ok := err.(types.ErrServiceClient); ok && (errsc.StatusCode == http.StatusNotFound) {
-			s.LoggingClient.Info(fmt.Sprintf("DeviceService %s doesn't exist, creating a new one", s.ServiceName))
-			id, err := s.edgexClients.DeviceServiceClient.Add(ctx, &newDeviceService)
-			if err != nil {
-				s.LoggingClient.Error(fmt.Sprintf("Failed to add Deviceservice %s: %v", s.ServiceName, err))
-				return err
+		if errors.Kind(err) == errors.KindEntityDoesNotExist {
+			s.LoggingClient.Infof("device service %s doesn't exist, creating a new one", newDeviceService.Name)
+			req := requests.AddDeviceServiceRequest{
```
Comment (cloudxxx8) [inline pkg/service/service.go:152]: use NewAddDeviceServiceRequest when it is available

File: pkg/service/service.go [hunk 24]
```diff
@@ -135,109 +131,64 @@ func (s *DeviceService) Stop(force bool) {
...
-			s.LoggingClient.Error(fmt.Sprintf("DeviceServicForName failed: %v", err))
+			s.LoggingClient.Errorf("failed to find device service %s", newDeviceService.Name)
 			return err
 		}
 	} else {
-		s.LoggingClient.Info(fmt.Sprintf("DeviceService %s exists, updating it", ds.Name))
-		err = s.edgexClients.DeviceServiceClient.Update(ctx, newDeviceService)
+		s.LoggingClient.Infof("device service %s exists, updating it", s.ServiceName)
+		req := requests.UpdateDeviceServiceRequest{
+			BaseRequest: commonDTO.BaseRequest{
```
Comment (cloudxxx8) [inline pkg/service/service.go:172]: use NewBaseRequest

File: pkg/service/service.go [hunk 25]
```diff
@@ -135,109 +131,64 @@ func (s *DeviceService) Stop(force bool) {
...
+				RequestId: uuid.New().String(),
+			},
+			Service: dtos.UpdateDeviceService{
```
Comment (cloudxxx8) [inline pkg/service/service.go:175]: add apiVersion

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]

File: pkg/models/manager.go [hunk 26]
```diff
@@ -0,0 +1,17 @@
+//
+// Copyright (C) 2021 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package models
+
+type Manager interface {
```
Comment (cloudxxx8) [inline pkg/models/manager.go:8]: Rename the interface or methods in v2 Manager model.
Manager is a general term.  If it is only for managing AutoEvent, it should be called AutoEventManager.
Or, we should rename the RestartForDevice to RestartAutoEventForDevice

File: internal/common/utils.go [hunk 27]
```diff
@@ -276,30 +37,55 @@ func GetUniqueOrigin() int64 {
...
 
-func FilterQueryParams(queryParams string, lc logger.LoggingClient) url.Values {
-	m, err := url.ParseQuery(queryParams)
+func UpdateLastConnected(name string, lc logger.LoggingClient, dc interfaces.DeviceClient) {
+	t := time.Now().UnixNano() / int64(time.Millisecond)
+	req := requests.UpdateDeviceRequest{
+		BaseRequest: common.BaseRequest{
+			RequestId: uuid.New().String(),
+		},
+		Device: dtos.UpdateDevice{
```
[the remark this answers is not in the document]
Comment (hahattan) [reply internal/common/utils.go:46]: `UpdateDevice` doesn't have `Versionable` right now 
https://github.com/edgexfoundry/go-mod-core-contracts/blob/6e13956c373f8385b66ccdf7b0d57d3383bd1502/v2/dtos/device.go#L36-L51

File: pkg/service/managedautoevents.go (hunk 2 above)
Comment (hahattan) [reply pkg/service/managedautoevents.go:61]: you're right we don't need to stop here.

File: internal/transformer/transform.go [hunk 28]
```diff
@@ -0,0 +1,140 @@
...
+
+	return
+}
+
+func commandValueToReading(cv *models.CommandValue, deviceName, profileName, mediaType, encoding string) dtos.BaseReading {
+	if encoding == "" {
+		encoding = models.DefaultFloatEncoding
+	}
+
+	reading := dtos.BaseReading{
```
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply internal/transformer/transform.go:109]: I opened a new issue for this
https://github.com/edgexfoundry/device-sdk-go/issues/788

**Review by cloudxxx8 — APPROVED** [review]
LGTM

**Pull request merged**